pkg/kubelet/status/status_manager.go

```diff
@@ -888,3 +891,32 @@ func NeedToReconcilePodReadiness(pod *v1.Pod) bool {
 	}
 	return false
 }
+
+func normalizeContainerResaon(pod *v1.Pod, status *v1.PodStatus) *v1.PodStatus {
+	// When pod was evicted, preempted, admitted failure or even terminated in response to imminent node shutdown,
+	// the pod became Failed, kubelet would set accurate reason for the pod.
+	// If the restart policy is Never and the pod becomes failed, kubelet would not set reason, that is not the case.
+	if status.Phase == v1.PodFailed && len(status.Reason) != 0 {
+		// update container reason
+		for i := range status.ContainerStatuses {
+			cstatus := &status.ContainerStatuses[i]
+			if cstatus.State.Terminated != nil {
+				cstatus.State.Terminated.Reason = status.Reason
+			} else if cstatus.State.Waiting != nil {
+				cstatus.State.Waiting.Reason = status.Reason
+			}
+		}
+
+		// update init container reason
+		for i := range status.InitContainerStatuses {
+			cstatus := &status.InitContainerStatuses[i]
+			if cstatus.State.Terminated != nil {
+				cstatus.State.Terminated.Reason = status.Reason
+			} else if cstatus.State.Waiting != nil {
+				cstatus.State.Waiting.Reason = status.Reason
+			}
+
+		}
```

Comment:
This part of code is kind of duplicate as above one for update the container reason, maybe could create a separate function and pass the container/initContainer status as variable to function to achieve the update.